UP Deduplication Design Documentation

JFisk42 opened the pull request:
This PR is for the UP Deduplication design. There are some outstanding research items which could affect minor pieces of the described design.

## Changes
- Added UP Deduplication Design Documentation

## Linked Issues
- Fixes #14103

## Outstanding TODOs 
_Should be resolved before merge_:  
- [x] Database processing efficiency investigation.
- [x] Is sender name collision a concern?
  - A: No, sender names are unique to an organization. 
- [x] Product Q: Are Azure Event logs needed for instances of duplicate detection?
  - A: From a platform perspective this is not needed
- [x] Engagement Q: Are Azure Event logs needed for instances of duplicate detection?
  - A: Engagement would like an Azure Event logged with parent_report_id, child_report_id, and sender org/name.
- [x] Update UP SRD with dedupe requirements/convert step changes as necessary.
  - A: This will be completed as part of the implementation.

arnejduranovic commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,197 @@
+Deduplication shall be performed at the item level (a FHIR Bundle) after conversion to FHIR happens. Each bundle will have key elements (see [Key FHIR Elements table](#key-fhir-elements)) converted to a string which is then hashed deterministically and stored in the database (item_lineage.item_hash). That hash will only be compared against other hashes from the same sender that were created withi ...
+
+#### Scope for Initial Implementation
+Deduplication will only be implemented for ORU_R01 messages from FULL_ELR topic senders. Covid Pipeline deduplication functionality should remain unchanged.
```
```suggestion
Deduplication will only be implemented for ORU_R01 messages from FULL_ELR topic senders. Covid Pipeline deduplication functionality shall remain unchanged.
```

arnejduranovic commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,197 @@
+
+# Universal Pipeline Deduplication
```
Consider adding a new section at the bottom that has the proposed convert step SRD changes so those can be reviewed as well. If these changes are only to the convert step, then I would copy the convert step requirements, amend them, and label each new or changed req with "(NEW)" tag.

JFisk42 replied:
Added Convert step changes under "Proposed Universal Pipeline SRD Additions", please review those when you have time.

Also noticed that there is some related language under 
```
Submission Step Business Logic Requirements 
    6.4. The submission step shall NOT accept the other custom HTTP Headers currently defined in RequestFunction.kt:
        6.4.4. AllowDuplicate – no clear requirements for this in UP at this time
```

This might also be a @jsutantio question. From my recollection we've been pretty adamant about not allowing senders to avoid this workflow on a per request basis. If that is still true then I would want to update this description to reflect that.

JFisk42 replied:
Spoke with @jsutantio after standup today and she confirmed that in the UP we do not want to enable this HTTP param which could override the sender settings for duplication. I have pushed a commit to add this section to proposed UP SRD changes:
```
6.4. [No Change] The submission step shall NOT accept the other custom HTTP Headers currently defined in RequestFunction.kt: 
    6.4.4. [UPDATED] AllowDuplicate – Not to be used for UP De-duplication
```

arnejduranovic commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,197 @@
+The first part of the Deduplication Workflow will take in the item’s Bundle and use the specified fields ([Key FHIR Elements table](#key-fhir-elements)) to generate a string which is then hashed. The current hashing implementation ([Report.getItemHashForRow()](https://github.com/CDCgov/prime-reportstream/blob/15648395efc2b60322d931bf88e0c2c5b6cc0371/prime-router/src/main/kotlin/Report.kt#L1287-L1 ...
+
+Technical Considerations:
+- This should be flexible enough that different senders, message types, and topics can implement different key fields.
```
How will this be achieved? I.e. Strategy design pattern? Something similar to how convert step does item [validation](https://github.com/CDCgov/prime-reportstream/blob/25f5e2bba34c67f4033c8ffaa67d2ee8406b83f5/prime-router/src/main/kotlin/validation)?

JFisk42 replied:
I've thought about this a couple of times and I keep coming back to yaml files (when you work at the ~~yaml~~hammer factory everything looks like a nail). Ultimately, the key fields are just a list of static fhirpaths. If you're simply looking for more specificity to address this consideration, I could prescribe a class which has the responsibility to reads in yaml files and handles the string concatenation. Then it could later be enhanced to take in a yaml filepath if we wanted to go down that road.

We could use something more OOP like a pattern if there was anything they were responsible for outside of reading in the specific paths. Off the top of my head the only example justification I can think of is something along the lines of "what if a sender/topic/message type needs to violate the core requirements of UP dedup". Ex: For X sender we only look at past month of results. For Y topic we want to compare against all senders.

JFisk42 replied:
TLDR: I've updated Technical Consideration with some technical description and a short mock of the proposed class structure (I believe this is in fact the Strategy design pattern but I am rusty on patterns in Java).

Actually, after mulling over this strategy again I don't like the idea of using yaml files to host fhirpath for the key fields. Storing the business logic for the key fields in Kotlin will avoid having to define yet another yaml schema and having to load/store them for the convert step. The likelihood of needing sender or topic specific key fields seems very low. Additionally, this will reduce steps in this part of the process. (There will no longer need to be a call to a fhirpath util method to extract values before concatenation of each field, instead extraction and concatenation can occur at the same time.

arnejduranovic commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,197 @@
+Technical Considerations:
+- This should be flexible enough that different senders, message types, and topics can implement different key fields.
+- The ordering of fields within a FHIR Bundle are **not** guaranteed. Before converting to a string to be hashed, the [populated key fields](#key-fhir-elements) should be extracted from the bundle and put in a static order (alphabetical etc.).
+- Sender id/name needs to be incorporated into pre-hashed string. (Dependent upon sql efficiency investigation noted below).
```
```suggestion
- Sender id/name shall be incorporated into the pre-hashed string. (Dependent upon sql efficiency investigation noted below).
```

arnejduranovic commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,197 @@
+    - (Comment from Michael: "For this to be performative you would need to be sure that this index can fit in memory of postgresql instead of having to go into disk.")
+
+##### Item Hash Comparison
+Hash comparison will be skipped if deduplication is disabled on that sender's settings. The generated item hash shall first be compared with items in the same report. (Note: At this part of the Convert Step, parallelization is not a blocker as the original items are [still within scope](https://github.com/CDCgov/prime-reportstream/blob/15648395efc2b60322d931bf88e0c2c5b6cc0371/prime-router/src/mai ...
```
Define the setting name

JFisk42 replied:
Somehow I previously missed that we should be able to entirely repurpose the covid pipeline sender setting for this. I've updated this document to reflect that, as well as a couple other docs that made reference to it in the context of the UP. https://github.com/CDCgov/prime-reportstream/pull/17143/commits/798d27495092457188697aa05e99a021d1bd3a9b

arnejduranovic commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,197 @@
+            "trackingIds": [
+                "849b3151-25f7-41f3-b19b-fd8ad47bae18"
+            ],
+            "message": "Report could not processed. Reason: Duplicate report was detected and removed.",
```
```suggestion
            "message": "Report could not processed. Reason: Duplicate report was detected and removed (no unique test results were found in the report.",
```

or some version of the above that sheds light on what a duplicate report actually is. This might be a good thing for @jsutantio to review.

JFisk42 replied:
Ah thanks for the callout. I should have also pointed out the content of the warning/error messages when asked this morning what @jsutantio should review here. Personally, I don't love what I came up with, but I am somewhat ambiguous on what the right vocab is for the submission history API audience.

arnejduranovic commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,197 @@
+
+<BR><img alt="Dedupe_Design.jpg" height="310" src="Dedupe_Design.jpg" width="450"/>
+
+- The deduplication workflow will happen at the very end of the convert step processing ([FHIRConverter.kt#L541](https://github.com/CDCgov/prime-reportstream/blob/c942a9a6f6be347d82196939e1cf677512af4a06/prime-router/src/main/kotlin/fhirengine/engine/FHIRConverter.kt#L541)).
```
According to your diagram it does not happen "at the very end"

JFisk42 replied:
I'll update the wording here. This is referring to the FHIRConverter method process() which does the everything listed before deduplication. After that method there are still FHIRConverter actions taken but they're more related to informational actions (updating the db, action logging, azure events, queueing, etc).

arnejduranovic commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,197 @@
+- The deduplication workflow will happen at the very end of the convert step processing ([FHIRConverter.kt#L541](https://github.com/CDCgov/prime-reportstream/blob/c942a9a6f6be347d82196939e1cf677512af4a06/prime-router/src/main/kotlin/fhirengine/engine/FHIRConverter.kt#L541)).
+  - Before the Deduplication Workflow, the Convert Step will:
+    - Convert to FHIR, if necessary
+    - De-batched items from the report
```
```suggestion
    - De-batch items from the report
```

arnejduranovic commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,197 @@
+  - Before the Deduplication Workflow, the Convert Step will:
+    - Convert to FHIR, if necessary
+    - De-batched items from the report
+    - Parse these items into bundles
```
```suggestion
    - Convert the items into FHIR bundles
```

arnejduranovic commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,197 @@
+    - Convert to FHIR, if necessary
+    - De-batched items from the report
+    - Parse these items into bundles
+    - Validate the items
```
This happens before converting, no?

JFisk42 replied:
Good catch. 
Looking over these lines a bit closer it appears that was the only thing out of order. I'll move it to before the convert.

arnejduranovic reviewed (commented):
Looks great! Please address TODOs before merging.

JFisk42 commented:
Key changes since last reviews:
* New DB recommendations based on discovery that only 5% of item_lineage.item_hash values are necessary for deduplication. (See "Other Updates")
* Added proposal for 2-step rollout. Talked with Jessica and having some time to gather item_hash data to analyze potential false positives would be a good safety check.
* Added UP SRD update proposals
* Documented existing sender setting for enabling duplicate detection
* Addition of Azure Events for engagement team
* Mock class design for containing key field business logic

stevewassnava commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,248 @@
+ There are three key differences between the deduplication designs. In the Universal Pipeline,
+- Deduplication will now happen during the Convert Step instead of the Receive Step.
+- Deduplication will only consider a hash of key fields instead of the entire submitted item.
+- Deduplication will only consider hashes from the same sender in the last year.
```
question: does this mean the calendar year or a year to date? I think this means to say within the past year to date (trying to think of a more concise way to send) - #2 below

_e.g., hash X sent 2/6/25, do we consider_
1. hashes from 2/6/24- 2/6/25 **OR**
2. hashes from 2024

JFisk42 replied:
Yeah this is a good point about the ambiguity. It may be better documented on the epic but I'll update the wording here. The answer is #1 here so I'll update it to be "from the same sender _within_ the last year" or could do something a slightly more verbose if that isn't sufficient either.

JFisk42 replied:
Added this now.

stevewassnava replied:
confirmed- update looks good- resolving

stevewassnava commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,248 @@
+  - If deemed necessary, sender specific key field classes could be created and then referenced in the sender settings with a new sender field.
+- The ordering of fields within a FHIR Bundle are **not** guaranteed. Before converting to a string to be hashed, the [populated key fields](#key-fhir-elements) should be extracted from the bundle and put in a static order (alphabetical etc.).
+- Sender id/name shall be incorporated into the pre-hashed string. (Dependent upon sql efficiency investigation noted below).
+- Investigation TODO (_To be removed and appropriate sections updated before merge_): Investigate the time efficiency of the following options:
```
is this still the case - to update before the merge of THIS PR, correct?

JFisk42 replied:
Correct, Arnej [added an index](https://github.com/CDCgov/prime-reportstream/pull/17279) to one of the columns on this table. Basically, using metabase to determine the db efficiency needs could not be done because all the queries were dying.

stevewassnava commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,248 @@
+  - Adding the sender id (or name) into the string to be hashed and allowing the column index to be the main SQL query parameter.
+
+##### Item Hash Comparison
+Hash comparison will be skipped if `sender.allowDuplicates` is set to true. Otherwise, the generated item hash shall first be compared with items in the same report. (Note: At this part of the Convert Step, parallelization is not a blocker as the original items are [still within scope](https://github.com/CDCgov/prime-reportstream/blob/15648395efc2b60322d931bf88e0c2c5b6cc0371/prime-router/src/main ...
```
would it be better to note the existing name of the function /variable/etc. in addition to the link to the specific line numbers? That is perfect for right now, but those references can easily get out of date when files are updated / altered in the future (and may not cause an update for this file)

JFisk42 replied:
I try to use permalinks for things like this (let me know if there are spots where I am not). GitHub can create permalinks to a line on any branch or commit to prevent against this very thing. (Actually... looking now, if you're on a branch the most recent commit will be what is referenced. I want to say that permalinks should also be able to refer to commits that are "wiped out" by things like a rebase/force push where that commit is no longer in history, but now that I tried to google the answer I didn't see an answer either way.)
![Screenshot 2025-02-06 at 11 30 43 AM](https://github.com/user-attachments/assets/ea727929-05c1-4386-9e61-cb1069caff28)

stevewassnava replied:
thank you Josh- will have to note this for my future reference - sounds good!

arnejduranovic commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,248 @@
+  - Long term change: Modify the index on item_lineage to a partial index, on only those values which are not null.
+    - Caveat: Barring a significant database wipe of out of date information, this would likely result in a significant _decrease_ in database efficiency.
+    - A hybrid plan would be to begin storing NULL values. After older values have been culled, and/or enough NULL values have accumulated in the table, then change the index to a partial index. 
+- Database: Investigation will need to happen on how to best purge out of date records on the item_lineage table. item_lineage.created_at will be used for this. This is outside the scope for this design/implementation.
```
If there is an index on created_at performance should not be affected that much if we just limit our select query by created_at?

JFisk42 replied:
Correct, and this was more aimed at the original notion of periodically purging the tables. I've removed this line as it isn't really relevant to the design here and added an "investigation results" bullet point under technical considerations with a brief summary of the db efficiency investigation and how created_at was relevant to that.

arnejduranovic commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,248 @@
+- Receiver Step: There is stubbed code to be removed here: The SumissionReceiver refers to the once theorized area where UP deduplication would be invoked. ([SubmissionReceiver.kt#L284-L292](https://github.com/CDCgov/prime-reportstream/blob/0c5e0b058e35e09786942f2c8b41c1d67a5b1d16/prime-router/src/main/kotlin/SubmissionReceiver.kt#L284-L292))
+- Database Changes: Several changes will need to take place to improve efficiency. **Over 90% of the values stored in item_lineage.item_hash are not relevant to either CP or UP deduplication.**
+  - Immediate change: Modify item_lineage.item_hash to accept NULL values.
+    - Reasoning: 90%+ of item_lineage table rows have no need to store the item_hash. Allowing null values will reduce updates to the index and indexing times for the table. 
```
Hmm, I wouldn't say they have NO reason to store these. They may not be relevant to deduplication but they do technically record the item in the DB long after it has been deleted from the blob store. This MAY be useful for debugging purposes but probably not... For example, someone says "I received this item from RS and its wrong" we could hash it and check against the DB if it is indeed what we sent or if it has been since modified

JFisk42 replied:
That's a fair point but I think this will have a positive effect on performance around this table, especially long term as more null values are added. I think pulling the report from blob storage would be better for this kind of scenario, but I'm not sure how long those stick around for.

JFisk42 commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,248 @@
+    - Apply sender schema transform
+    - Make a report object with item lineage
+      - **This will now include the newly generated (or null) hash**
+    - Create a route event
```
Ope, this was referring to the convert step's existing azure events. But it made me realize I hadn't named the new azure event(s). I've added that under the item hash comparison section where the azure event is described.

JFisk42 commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,248 @@
+  - During the Deduplication Workflow, the Convert Step will:
+    - Generate a hash for the item
+    - Compare this hash against those batched in the same submitted report
+    - Compare this hash with others stored in the database
```
Added.

JFisk42 commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,248 @@
+            "trackingIds": [
+                "849b3151-25f7-41f3-b19b-fd8ad47bae18"
+            ],
+            "message": "Duplicate message was detected and removed.",
```
Now that's an interesting question. That should be relatively easy to do but I'll let @jsutantio weigh in on if she wants that or any other details here.

JFisk42 commented on prime-router/docs/design/design/deduplication/deduplication.md:
```diff
@@ -0,0 +1,248 @@
+
+Technical Considerations:
+- This should be flexible enough that different senders, message types, or even topics can implement different key fields.
+  - This will be achieved through a new base class which shall own the logic for concatenating strings values in a map. Each set of new key fields would implement its own class which owns the logic knowing how to retrieve the necessary FHIR values. [See: Example Key Fields Class](#example-key-fields-file-structure) 
```
Added a suggestion to use `LinkedHashMap` for this reason. I think there's probably some other good ways to get the fields into an ordered string so I won't go more strict than suggesting this.
